File: functions-and-parameters/FunctionsParameters.cpp
#include <functional>
#include <iostream>
#include <string>

std::string greet(const std::string& name, const std::string& prefix = "Hello") {
    return prefix + ", " + name + "!";
}

// Overloading example
int add(int a) {
    return add(a, 0);
}

int add(int a, int b) {
    return a + b;
}

int apply_operation(int a, int b, const std::function<int(int, int)>& operation) {
    return operation(a, b);
}

int main() {
    std::cout << greet("Nazmul") << "\n";

    // C++ has no named parameters; pass by position.
    std::cout << greet("Nazmul", "Hi") << "\n";

    std::cout << add(10) << "\n";
    std::cout << add(10, 5) << "\n";

    // First-class function: a lambda can be stored in a variable and passed to another function.
    auto multiply = [](int x, int y) { return x * y; };
    std::cout << apply_operation(4, 3, multiply) << "\n";

    return 0;
}

File: functions-and-parameters/functions_parameters.py
def greet(name: str, prefix: str = "Hello") -> str:
    return f"{prefix}, {name}!"


def add(a: int, b: int = 0) -> int:
    return a + b


def apply_operation(a: int, b: int, operation):
    return operation(a, b)


def main() -> None:
    # Default parameter
    print(greet("Nazmul"))

    # Named parameter
    print(greet(name="Nazmul", prefix="Hi"))

    # Python has no classic overloading; defaults often replace it
    print(add(10))
    print(add(10, 5))

    # First-class function: a function can be stored in a variable and passed to another function.
    multiply = lambda x, y: x * y
    print(apply_operation(4, 3, multiply))


if __name__ == "__main__":
    main()
